code/parachain/runtime/dali/src/lib.rs

```diff
@@ -1542,6 +1542,45 @@ impl_runtime_apis! {
 		}
 	}
 
+	impl cosmwasm_runtime_api::CosmwasmRuntimeApi<Block, AccountId, CurrencyId, Balance, Vec<u8>> for Runtime {
+		fn query(
+			contract: AccountId,
+			gas: u64,
+			query_request: Vec<u8>,
+		) -> Result<Vec<u8>, Vec<u8>>{
+			match cosmwasm::query::<Runtime>(
+				contract,
+				gas,
+				query_request,
+			) {
+				Ok(response) => Ok(response.0),
+				Err(err) => Err(alloc::format!("{:?}", err).into_bytes())
+			}
+		}
+
+		fn instantiate(
+			instantiator: AccountId,
+			code_id: u64,
+			salt: Vec<u8>,
+			admin: Option<AccountId>,
+			label: Vec<u8>,
+			funds: BTreeMap<CurrencyId, (Balance, bool)>,
+			gas: u64,
+			message: Vec<u8>,
+		) -> Result<AccountId, Vec<u8>> {
+			cosmwasm::instantiate::<Runtime>(
+				instantiator,
+				code_id,
+				salt,
+				admin,
+				label,
+				funds,
+				gas,
+				message
+			).map_err(|err| alloc::format!("{:?}", err).into_bytes())
```

Comment:
sp_std::format?

Reply:
they say that it is preferable to use `format!` instead in their docs. https://docs.rs/sp-std/latest/sp_std/fmt/fn.format.html

Reply:
I believe dzmitry was referring to using `sp_std::format!` instead of `alloc::format!`, but it doesn't really matter since sp_std is just reexports of core/alloc/std.

Reply:
yes. just we use sp_std a lot and it maps to whatever it maps. so any part of code is portable.

Reply:
```
Module [sp_std](https://docs.rs/sp-std/latest/sp_std/index.html)::[fmt](https://docs.rs/sp-std/latest/sp_std/fmt/index.html#)Copy item path
1.0.0[[−]](javascript:void(0))[[src]](https://doc.rust-lang.org/nightly/src/alloc/lib.rs.html#196)
Utilities for formatting and printing Strings.

This module contains the runtime support for the [format!](https://doc.rust-lang.org/nightly/alloc/macro.format.html) syntax extension. This macro is implemented in the compiler to emit calls to this module in order to format arguments at runtime into strings.

[Usage](https://docs.rs/sp-std/latest/sp_std/fmt/index.html#usage)
The [format!](https://doc.rust-lang.org/nightly/alloc/macro.format.html) macro is intended to be familiar to those coming from C’s printf/fprintf functions or Python’s str.format function.

Some examples of the [format!](https://doc.rust-lang.org/nightly/alloc/macro.format.html) extension are:

format!("Hello");                 // => "Hello"
format!("Hello, {}!", "world");   // => "Hello, world!"
format!("The number is {}", 1);   // => "The number is 1"
format!("{:?}", (3, 4));          // => "(3, 4)"
format!("{value}", value=4);      // => "4"
let people = "Rustaceans";
format!("Hello {people}!");       // => "Hello Rustaceans!"
format!("{} {}", 1, 2);           // => "1 2"
format!("{:04}", 42);             // => "0042" with leading zeros
format!("{:#?}", (100, 200));     // => "(
                                  //       100,
                                  //       200,
                                  //     )"
From these, you can see that the first argument is a format string. It is required by the compiler for this to be a string literal; it cannot be a variable passed in (in order to perform validity checking). The compiler will then parse the format string and determine if the list of arguments provided is suitable to pass to this format string.

To convert a single value to a string, use the [to_string](https://doc.rust-lang.org/nightly/alloc/string/trait.ToString.html#tymethod.to_string) method. This will use the [Display](https://docs.rs/sp-std/latest/sp_std/fmt/trait.Display.html) formatting trait.
```